lib/RestructureCFG/InlineDispatcherSwitch.cpp

```diff
@@ -968,12 +980,44 @@ static RecursiveCoroutine<ASTNode *> simplifySwitchBreakImpl(ASTNode *Node) {
 
         if (llvm::isa<SwitchBreakNode>(Case)) {
           ToRemoveCaseIndex.push_back(Index);
+
+          // If we are removing an atomic `Label`, we save it for later removal
+          // of the associated `SetNode`. We do not remove the `SetNode`s
+          // associated to multiple `Label`s `case`s, because that would not be
+          // correct. The removal of the associated `SetNode` should be
+          // performed only in correspondence of the `case`s containing the
+          // atomic `Label`.
+          if (LabelSet.size() == 1) {
+            RemovedAtomicLabels.insert(*LabelSet.begin());
+          }
         }
       }
 
       for (auto ToRemoveCase : llvm::reverse(ToRemoveCaseIndex)) {
         Switch->removeCaseN(ToRemoveCase);
       }
+
+      // In case we are simplifying the `SwitchBreak` associated to an exit
+      // dispatcher, we also need to remove the corresponding `SetNode`s
+      using DispatcherKind = typename SwitchNode::DispatcherKind;
+      if (not Switch->getCondition()
+          and Switch->getDispatcherKind() == DispatcherKind::DK_Exit) {
+
+        // Precompute how many `SetNode`s are present in the related `Scs`
+        auto &[RelatedLoop, RemoveSetNode] = LoopDispatcherMap.at(Switch);
+        SetNodeCounterMap SetCounterMap;
+        countSetNodeInLoop(RelatedLoop->getBody(), SetCounterMap);
+
+        // Remove the `SetNode`s associated to the removed `Label`s
+        for (auto &Label : RemovedAtomicLabels) {
+          auto &Sets = SetCounterMap.at(Label);
+          for (auto *Set : Sets) {
+
+            // To perform the removal, we reuse the `addToDispatcherSet`
+            addToDispatcherSet(AST, RelatedLoop->getBody(), nullptr, Set, true);
```

Comment:
Please make a dedicated directive for this. Using `addToDispatcherSet` with `nullptr` leaks internal details and is slightly confusing if you ignore those details.

Reply:
Right, I created a new wrapper to hide the internal working details [here](https://github.com/revng/revng-c/blob/4c08f81ad95989ec412c54a03f1e80560eef7a0a/lib/RestructureCFG/InlineDispatcherSwitch.cpp#L226)